core/templates/pages/exploration-editor-page/services/version-history.service.ts

```diff
@@ -0,0 +1,314 @@
+// Copyright 2022 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service for the exploration states version history.
+ */
+
+import { Injectable } from '@angular/core';
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { ExplorationMetadata } from 'domain/exploration/ExplorationMetadataObjectFactory';
+import { State } from 'domain/state/StateObjectFactory';
+
+export interface StateDiffData {
+  oldState: State | null;
+  newState: State | null;
+  oldVersionNumber: number | null;
+  newVersionNumber: number | null;
+  committerUsername: string;
+}
+
+export interface MetadataDiffData {
+  oldMetadata: ExplorationMetadata | null;
+  newMetadata: ExplorationMetadata | null;
+  oldVersionNumber: number | null;
+  newVersionNumber: number | null;
+  committerUsername: string;
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class VersionHistoryService {
+  // The following value is null when the service is not initialized i.e.
+  // the user hasn't opened any exploration editor page yet.
+  latestVersionOfExploration: number | null = null;
+
+  // The following four arrays namely fetchedStateVersionNumbers,
+  // fetchedMetadataVersionNumbers, fetchedStateData, fetchedMetadata can
+  // contain only their last element as null and no other element of these
+  // arrays will be none. The element null marks the end of the version
+  // history of a particular state or exploration metadata. When we keep
+  // fetching the 'previously edited versions' of a state or exploration
+  // metadata, the backend returns the version number as 'None' when we
+  // reach the end of the version history.
+  fetchedStateVersionNumbers: (number | null) [] = [];
+  fetchedMetadataVersionNumbers: (number | null) [] = [];
+  fetchedStateData: (State | null) [] = [];
+  fetchedMetadata: (ExplorationMetadata | null) [] = [];
+
+  fetchedCommitterUsernames: string[] = [];
+  currentPositionInStateVersionHistoryList: number = 0;
```

Comment:
Can you add comments explaining what these variables mean? I don't understand what they mean and that's making it hard for me to evaluate whether the use of 0 is sensible or not.

If 0 is the actually-correct value (it actually logically represents the current position) then this should be fine. But if you're using it as a "sentinel" value just to satisfy the typing (and it could be replaced by -1, say, without consequence), then that's not fine and we need to discuss this further.

Reply:
It represents the current position in the version history list of a particular state or metadata when the user is navigating through the list. Earlier, I was using null to represent when the service was not initialized. But later on, I changed it to zero to satisfy the typing.

But, I think null or -1 will be a conceptually good sentinel value because they can represent that the service has not been initialised yet. After the initialisation, it will be set to zero to represent the first element.